### PR TITLE
:seedling: reconciler/cache/reconciler: simplify and generalize

### PR DESCRIPTION
**Pull request by sttts**
Generalized to host more resources more easily and simplified whole flow.
Generalized to host more resources more easily and simplified whole flow.
Generalized to host more resources more easily and simplified whole flow.
Generalized to host more resources more easily and simplified whole flow.
Generalized to host more resources more easily and simplified whole flow.

### CODE REVIEW

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 1]
```diff
@@ -24,194 +24,178 @@ import (
...
+	globalExists := !errors.IsNotFound(err)
 
-	var unstructuredCacheObject *unstructured.Unstructured
-	var unstructuredLocalObject *unstructured.Unstructured
-	if cacheObject != nil {
-		unstructuredCacheObject, err = toUnstructured(cacheObject)
-		if err != nil {
-			return err
+	// local is gone or being deleted. Delete in cache.
+	if !localExists || !localCopy.GetDeletionTimestamp().IsZero() {
```
Comment (fgiloux) [inline pkg/reconciler/cache/replication/replication_reconcile.go:153]: How does it play with finalizers? Doesn't it remove the object from the cache when finalizers may be holding the local copy? Do we want that?

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 2]
```diff
@@ -24,194 +24,178 @@ import (
...
-	if localObject != nil {
-		unstructuredLocalObject, err = toUnstructured(localObject)
-		if err != nil {
-			return err
+
+		// Object doesn't exist anymore, delete it from the global cache.
+		logger.V(1).WithValues("cluster", clusterName, "namespace", ns, "name", name).Info("Deleting object from global cache")
+		if err := r.deleteObject(ctx, clusterName, ns, name); err != nil && !errors.IsNotFound(err) {
+			runtime.HandleError(err)
+			return nil
```
Comment (fgiloux) [inline pkg/reconciler/cache/replication/replication_reconcile.go:162]: no retry?

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 3]
```diff
@@ -24,194 +24,178 @@ import (
...
 	}
-	if cluster.Empty() && localObject != nil {
-		metadata, err := meta.Accessor(localObject)
-		if err != nil {
-			return err
+
+	// local exists, global doesn't. Create in cache.
+	if !globalExists {
+		// TODO: in the future the original RV will have to be stored in an annotation (?)
+		// so that the clients that need to modify the original/local object can do i
```
Comment (fgiloux) [inline pkg/reconciler/cache/replication/replication_reconcile.go:170]: nit: "can do it"

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 4]
```diff
@@ -24,194 +24,178 @@ import (
...
+	createObject func(ctx context.Context, cluster logicalcluster.Name, obj *unstructured.Unstructured) (*unstructured.Unstructured, error)
+	updateObject func(ctx context.Context, cluster logicalcluster.Name, obj *unstructured.Unstructured) (*unstructured.Unstructured, error)
+	deleteObject func(ctx context.Context, cluster logicalcluster.Name, ns, name string) error
+}
+
+// reconcile makes sure that the object under the given key from the local shard is replicated to the cache server.
 // the replication function handles the following cases:
-//  1. creation of the object in the cache server when the cached object is not found by retrieveLocalObject
-//  2. deletion of the object from the cache server when the original/local object was removed OR was not found by retrieveLocalObject
+//  1. creation of the object in the cache server when the cached object is not found by getLocalCopy
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:126]: yes

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:153]: we patched the cache server to ignore `Finalizers`

Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:153]: very good question. So we should keep updating it, but filter the deletion timestsamp?

Does the cache-server ignore or follow finalizers?

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:126]: yes, we create an obj when it wasn't found in the globalCache

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 1 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:153]: removing an object is enough

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 2 above)
Comment (p0lyn0mial) [reply pkg/reconciler/cache/replication/replication_reconcile.go:162]: oh, right we should retry

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:126]: done

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 1 above)
Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:153]: follow-up

File: pkg/reconciler/cache/replication/replication_reconcile.go [hunk 5]
```diff
@@ -24,194 +24,178 @@ import (
...
-		unstructuredCacheObject.SetKind(gvk.Kind)
-		unstructuredCacheObject.SetAPIVersion(gvr.GroupVersion().String())
-	}
-	if localObject != nil {
-		unstructuredLocalObject, err = toUnstructured(localObject)
-		if err != nil {
-			return err
+
+		// Object doesn't exist anymore, delete it from the global cache.
+		logger.V(1).WithValues("cluster", clusterName, "namespace", ns, "name", name).Info("Deleting object from global cache")
```
[the remark this answers is not in the document]
Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:159]: done

File: pkg/reconciler/cache/replication/replication_reconcile.go (hunk 2 above)
Comment (sttts) [reply pkg/reconciler/cache/replication/replication_reconcile.go:162]: done

**Pull request merged**